providers/apis/defi/osmosis/price_fetcher.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	if !api.Enabled {
+		return nil, fmt.Errorf("api is not enabled")
+	}
+
+	if apiMetrics == nil {
+		return nil, fmt.Errorf("metrics cannot be nil")
+	}
+
+	return &APIPriceFetcher{
+		api:               api,
+		client:            client,
+		logger:            logger.With(zap.String("fetcher", Name)),
+		metaDataPerTicker: newMetadataCache(),
+	}, nil
+}
+
+// Fetch fetches prices from the osmosis API for the given currency-pairs. Specifically
+// for each currency-pair,
+//   - Query the spot price.
+func (pf *APIPriceFetcher) Fetch(
+	ctx context.Context,
+	tickers []oracletypes.ProviderTicker,
+) oracletypes.PriceResponse {
+	resolved := make(oracletypes.ResolvedPrices)
+	unresolved := make(oracletypes.UnResolvedPrices)
+
+	wg := sync.WaitGroup{}
+	unresolvedMtx := sync.Mutex{}
+	resolveMtx := sync.Mutex{}
+	wg.Add(len(tickers))
+
+	// setup callbacks for writing to maps in parallel
+	unresolvedTickerCallback := func(ticker oracletypes.ProviderTicker, err providertypes.ErrorWithCode) {
+		unresolvedMtx.Lock()
+		defer unresolvedMtx.Unlock()
+		unresolved[ticker] = providertypes.UnresolvedResult{
+			ErrorWithCode: err,
+		}
+	}
+
+	resolvedTickerCallback := func(ticker oracletypes.ProviderTicker, price *big.Float) {
+		resolveMtx.Lock()
+		defer resolveMtx.Unlock()
+		resolved[ticker] = oracletypes.NewPriceResult(price, time.Now().UTC())
+	}
+
+	pf.logger.Info("fetching for tickers", zap.Any("tickers", tickers))
+
+	// make sure metadata cache is set properly
+	for _, ticker := range tickers {
+		_, found := pf.metaDataPerTicker.getMetadataPerTicker(ticker)
+		if !found {
+			_, err := pf.metaDataPerTicker.updateMetaDataCache(ticker)
+			if err != nil {
+				pf.logger.Debug("failed to update metadata cache", zap.Error(err))
+			}
+		}
+	}
+
+	for _, ticker := range tickers {
+		go func(ticker oracletypes.ProviderTicker) {
+			var err error
+			defer wg.Done()
+
+			metadata, found := pf.metaDataPerTicker.getMetadataPerTicker(ticker)
+			if !found {
+				unresolvedTickerCallback(ticker, providertypes.NewErrorWithCode(
+					NoOsmosisMetadataForTickerError(ticker.String()),
+					providertypes.ErrorTickerMetadataNotFound,
+				))
+
+				return
+			}
+
+			callCtx, cancel := context.WithTimeout(ctx, pf.api.Timeout)
+			defer cancel()
+
+			resp, err := pf.client.SpotPrice(callCtx,
+				metadata.PoolID,
+				metadata.BaseTokenDenom,
+				metadata.QuoteTokenDenom,
+			)
+			if err != nil {
+				unresolvedTickerCallback(ticker, providertypes.NewErrorWithCode(
+					err,
+					providertypes.ErrorAPIGeneral,
+				))
+
+				return
+			}
+
+			price, err := calculatePrice(resp)
+			if err != nil {
+				pf.logger.Error("failed parse spot price response", zap.Error(err))
+
+				unresolvedTickerCallback(ticker, providertypes.NewErrorWithCode(
+					err,
+					providertypes.ErrorFailedToParsePrice,
+				))
+
+				return
+			}
+
+			resolvedTickerCallback(ticker, price)
+		}(ticker)
+	}
+
+	wg.Wait()
```

Comment:
this should ideally be buffering a maximum number of concurrent requests otherwise we will certainly be rate limited.

Reply:
Now using an errgroup that limits the number of goroutines to `MaxQueries`